File: src/qq_group_ops/services/moderation.py
"""规则引擎。

当前实现只依赖标准库，方便 Phase 0/1 在没有安装任何第三方依赖时运行测试。
"""

from __future__ import annotations

import re
from collections.abc import Iterable
from dataclasses import dataclass

from qq_group_ops.core.enums import ModerationAction, RiskLevel
from qq_group_ops.core.models import RuleMatch

_ACTION_PRIORITY: dict[ModerationAction, int] = {
    ModerationAction.ALLOW: 0,
    ModerationAction.WARN: 1,
    ModerationAction.REVIEW: 2,
    ModerationAction.RECALL: 3,
    ModerationAction.MUTE: 4,
    ModerationAction.KICK: 5,
}


@dataclass(frozen=True, slots=True)
class ModerationRule:
    """一条可配置的审核规则。"""

    rule_id: str
    pattern: str
    action: ModerationAction = ModerationAction.WARN
    reason: str = "命中规则"
    risk: RiskLevel = RiskLevel.MEDIUM
    is_regex: bool = False
    enabled: bool = True
    case_sensitive: bool = False


class RuleEngine:
    """按规则逐条匹配消息内容。"""

    def __init__(self, rules: Iterable[ModerationRule] | None = None) -> None:
        self._rules = tuple(rules or ())
        self._compiled: dict[str, re.Pattern[str]] = {}
        for rule in self._rules:
            if rule.is_regex:
                flags = 0 if rule.case_sensitive else re.IGNORECASE
                try:
                    self._compiled[rule.rule_id] = re.compile(rule.pattern, flags)
                except re.error as exc:
                    raise ValueError(
                        f"invalid regex for rule {rule.rule_id!r}: {rule.pattern!r}"
                    ) from exc

    @property
    def rules(self) -> tuple[ModerationRule, ...]:
        return self._rules

    def evaluate(self, content: str) -> tuple[RuleMatch, ...]:
        """返回所有命中的规则。"""

        matches: list[RuleMatch] = []
        for rule in self._rules:
            if not rule.enabled:
                continue
            if not self._matches(rule, content):
                continue
            matches.append(
                RuleMatch(
                    rule_id=rule.rule_id,
                    pattern=rule.pattern,
                    action=rule.action,
                    reason=rule.reason,
                    risk=rule.risk,
                )
            )
        return tuple(matches)

    def highest_action(self, content: str) -> ModerationAction:
        """返回命中规则中优先级最高的动作。"""

        matches = self.evaluate(content)
        if not matches:
            return ModerationAction.ALLOW
        return max(matches, key=lambda match: _ACTION_PRIORITY[match.action]).action

    def _matches(self, rule: ModerationRule, content: str) -> bool:
        if rule.is_regex:
            return self._compiled[rule.rule_id].search(content) is not None
        if rule.case_sensitive:
            return rule.pattern in content
        return rule.pattern.lower() in content.lower()

    @classmethod
    def from_keywords(
        cls,
        keywords: Iterable[str],
        *,
        action: ModerationAction = ModerationAction.WARN,
        risk: RiskLevel = RiskLevel.MEDIUM,
    ) -> RuleEngine:
        """从关键词列表快速构建规则引擎。"""

        rules = [
            ModerationRule(
                rule_id=f"keyword:{keyword}",
                pattern=keyword,
                action=action,
                reason=f"命中关键词：{keyword}",
                risk=risk,
            )
            for keyword in keywords
            if keyword
        ]
        return cls(rules)

File: src/services/moderation.ts
import {
  ModerationAction,
  RiskLevel,
} from "../core/enums.js";
import type { RuleMatch } from "../core/models.js";

const ACTION_PRIORITY: Record<ModerationAction, number> = {
  [ModerationAction.Allow]: 0,
  [ModerationAction.Warn]: 1,
  [ModerationAction.Review]: 2,
  [ModerationAction.Recall]: 3,
  [ModerationAction.Mute]: 4,
  [ModerationAction.Kick]: 5,
};

export interface ModerationRule {
  ruleId: string;
  pattern: string;
  action?: ModerationAction;
  reason?: string;
  risk?: RiskLevel;
  isRegex?: boolean;
  enabled?: boolean;
  caseSensitive?: boolean;
}

interface ResolvedRule {
  ruleId: string;
  pattern: string;
  action: ModerationAction;
  reason: string;
  risk: RiskLevel;
  isRegex: boolean;
  enabled: boolean;
  caseSensitive: boolean;
}

export class RuleEngine {
  private readonly resolvedRules: readonly ResolvedRule[];
  private readonly compiled = new Map<string, RegExp>();

  public constructor(rules: readonly ModerationRule[] = []) {
    this.resolvedRules = rules.map((rule) => ({
      ruleId: rule.ruleId,
      pattern: rule.pattern,
      action: rule.action ?? ModerationAction.Warn,
      reason: rule.reason ?? "命中规则",
      risk: rule.risk ?? RiskLevel.Medium,
      isRegex: rule.isRegex ?? false,
      enabled: rule.enabled ?? true,
      caseSensitive: rule.caseSensitive ?? false,
    }));

    for (const rule of this.resolvedRules) {
      if (!rule.isRegex) {
        continue;
      }
      try {
        this.compiled.set(
          rule.ruleId,
          new RegExp(rule.pattern, rule.caseSensitive ? "" : "i"),
        );
      } catch (error) {
        throw new Error(
          `invalid regex for rule ${JSON.stringify(rule.ruleId)}: ${JSON.stringify(rule.pattern)}`,
          { cause: error },
        );
      }
    }
  }

  public get rules(): readonly ResolvedRule[] {
    return this.resolvedRules;
  }

  public evaluate(content: string): RuleMatch[] {
    const matches: RuleMatch[] = [];
    for (const rule of this.resolvedRules) {
      if (!rule.enabled || !this.matches(rule, content)) {
        continue;
      }
      matches.push({
        ruleId: rule.ruleId,
        pattern: rule.pattern,
        action: rule.action,
        reason: rule.reason,
        risk: rule.risk,
      });
    }
    return matches;
  }

  public highestAction(content: string): ModerationAction {
    const matches = this.evaluate(content);
    if (matches.length === 0) {
      return ModerationAction.Allow;
    }
    return matches.reduce((highest, match) =>
      ACTION_PRIORITY[match.action] > ACTION_PRIORITY[highest.action] ? match : highest,
    ).action;
  }

  public static fromKeywords(
    keywords: readonly string[],
    options: { action?: ModerationAction; risk?: RiskLevel } = {},
  ): RuleEngine {
    return new RuleEngine(
      keywords
        .filter((keyword) => keyword.length > 0)
        .map((keyword) => ({
          ruleId: `keyword:${keyword}`,
          pattern: keyword,
          action: options.action ?? ModerationAction.Warn,
          reason: `命中关键词：${keyword}`,
          risk: options.risk ?? RiskLevel.Medium,
        })),
    );
  }

  /**
   * §B1：把群配置里的正则规则编译成规则集。
   *
   * 单条非法正则不会让整个引擎抛错（配置可能来自旧数据 / 手工改库），只跳过该条；
   * `/rules set|add regex` 在写入前已经用 `requireValidRegex` 校验过。
   */
  public static fromRegex(
    patterns: readonly string[],
    options: { action?: ModerationAction; risk?: RiskLevel } = {},
  ): RuleEngine {
    const valid: ModerationRule[] = [];
    for (const raw of patterns) {
      const pattern = raw.trim();
      if (pattern.length === 0) {
        continue;
      }
      try {
        new RegExp(pattern, "u");
      } catch {
        continue;
      }
      valid.push({
        ruleId: `regex:${pattern}`,
        pattern,
        action: options.action ?? ModerationAction.Warn,
        reason: `命中正则：${pattern}`,
        risk: options.risk ?? RiskLevel.Medium,
        isRegex: true,
      });
    }
    return new RuleEngine(valid);
  }

  private matches(rule: ResolvedRule, content: string): boolean {
    if (rule.isRegex) {
      return this.compiled.get(rule.ruleId)?.test(content) ?? false;
    }
    if (rule.caseSensitive) {
      return content.includes(rule.pattern);
    }
    return content.toLowerCase().includes(rule.pattern.toLowerCase());
  }
}
